src/main/scala/uk/gov/nationalarchives/omega/api/services/ApiService.scala

```diff
@@ -0,0 +1,219 @@
+import pureconfig.generic.auto._
+import pureconfig.{ ConfigObjectSource, ConfigSource }
+import uk.gov.nationalarchives.omega.api.common.ServiceArgs
+import uk.gov.nationalarchives.omega.api.conf.ServiceConfig
+import uk.gov.nationalarchives.omega.api.connectors.JmsConnector
+import uk.gov.nationalarchives.omega.api.services.LocalMessageStore.PersistentMessageId
+import uk.gov.nationalarchives.omega.api.services.ServiceState.{ Started, Starting, Stopped, Stopping }
+
+import java.nio.file.{ Files, Paths }
+import java.util.concurrent.atomic.AtomicReference
+
+class ApiService(val config: ServiceConfig) {
+
+  implicit val loggerFactory: LoggerFactory[IO] = Slf4jFactory[IO]
+  implicit val logger: SelfAwareStructuredLogger[IO] = LoggerFactory[IO].getLogger
+
+  private var errorOnInvalidServiceState = false
+  private val state = new AtomicReference[ServiceState](Stopped)
+
+  private val localMessageStore = {
+    val messageStoreFolder = Paths.get(config.tempMessageDir)
+    Files.createDirectories(messageStoreFolder)
+    new LocalMessageStore(messageStoreFolder)
+  }
+
+  def start: IO[ExitCode] = {
+    // check if we can start the service
+    if (!switchState(Stopped, Starting)(config.errorOnInvalidServiceState)) {
+      return IO.pure(ExitCode.Error)
+    }
+
+    // copy config option to local state
+    this.errorOnInvalidServiceState = config.errorOnInvalidServiceState
+
+    try {
+      // attempt startup
+      val ec = doStart(config)
+      // switch to the started state
+      switchState(Starting, Started)(errorOnInvalidServiceState)
+      ec
+    } catch {
+      case e: Throwable =>
+        // we are in some inconsistent state as startup failed, so we must attempt shutdown
+        try {
+          doStop()
+          switchState(Starting, Stopped)(errorOnInvalidServiceState)
+          IO.pure(ExitCode.Error)
+        } finally
+          println(e.getMessage)
+    }
+
+  }
+
+  @throws[IllegalStateException]
+  def stop(): Unit = {
+    // check if we can stop the service
+    if (!switchState(Started, Stopping)(errorOnInvalidServiceState)) {
+      return
+    }
+
+    // attempt shutdown
+    doStop()
+    switchState(Stopping, Stopped)(errorOnInvalidServiceState)
+    () // Explicitly return unit
+  }
+
+  @throws[IllegalStateException]
+  private def switchState(from: ServiceState, to: ServiceState)(errorOnInvalidServiceState: Boolean): Boolean = {
+    val switched = state.compareAndSet(from, to)
+    if (!switched) {
+      val msg = s"Unable to switch ApiService state from: $from, to: $to."
+      if (errorOnInvalidServiceState) {
+        throw new IllegalStateException(msg)
+      } else {
+        logger.warn(msg)
+      }
+    }
+    switched
+  }
+
+  private def doStart(config: ServiceConfig): IO[ExitCode] = {
+
+    // TODO(AR) - one client, how to ack a consumer message after local persistence and then process it, and then produce a response
```

Comment:
I assume these are from the prototype. Are they all still relevant?

Reply:
These are @adamretter's TODO comments so I think we should leave them as an aide-mémoire for him.

Reply:
OK.